File: patterns/testing/arrange-act-assert.py
"""
Arrange-Act-Assert Pattern - Python

Three-phase test structure:
- Arrange: Set up test data and conditions
- Act: Execute the code under test
- Assert: Verify the results
"""

import unittest


# Example: Calculator
class Calculator:
    def add(self, a, b):
        return a + b
    
    def subtract(self, a, b):
        return a - b
    
    def multiply(self, a, b):
        return a * b


# Test examples using AAA pattern
class TestCalculator(unittest.TestCase):
    def test_add_should_return_sum_of_two_numbers(self):
        # Arrange
        calculator = Calculator()
        a = 5
        b = 3
        
        # Act
        result = calculator.add(a, b)
        
        # Assert
        self.assertEqual(result, 8)
    
    def test_subtract_should_return_difference(self):
        # Arrange
        calculator = Calculator()
        
        # Act
        result = calculator.subtract(10, 4)
        
        # Assert
        self.assertEqual(result, 6)


# Example usage
if __name__ == "__main__":
    print("=== Arrange-Act-Assert Pattern ===\n")
    unittest.main(verbosity=2)

File: patterns/testing/arrange-act-assert.js
/**
 * Arrange-Act-Assert Pattern - JavaScript
 * 
 * Three-phase test structure:
 * - Arrange: Set up test data and conditions
 * - Act: Execute the code under test
 * - Assert: Verify the results
 */

// Example test framework structure
class TestFramework {
    constructor() {
        this.tests = [];
    }
    
    test(name, fn) {
        this.tests.push({ name, fn });
    }
    
    async run() {
        for (const test of this.tests) {
            try {
                await test.fn();
                console.log(`✓ ${test.name}`);
            } catch (error) {
                console.log(`✗ ${test.name}: ${error.message}`);
            }
        }
    }
}

// Example: Calculator
class Calculator {
    add(a, b) {
        return a + b;
    }
    
    subtract(a, b) {
        return a - b;
    }
    
    multiply(a, b) {
        return a * b;
    }
}

// Test examples using AAA pattern
const framework = new TestFramework();

framework.test('Calculator.add should return sum of two numbers', () => {
    // Arrange
    const calculator = new Calculator();
    const a = 5;
    const b = 3;
    
    // Act
    const result = calculator.add(a, b);
    
    // Assert
    if (result !== 8) {
        throw new Error(`Expected 8, got ${result}`);
    }
});

framework.test('Calculator.subtract should return difference', () => {
    // Arrange
    const calculator = new Calculator();
    
    // Act
    const result = calculator.subtract(10, 4);
    
    // Assert
    if (result !== 6) {
        throw new Error(`Expected 6, got ${result}`);
    }
});

// Example usage
console.log('=== Arrange-Act-Assert Pattern ===\n');
framework.run();

module.exports = { TestFramework, Calculator };

